firewood/src/merkle/stream.rs

```diff
@@ -396,6 +396,140 @@ impl<'a, S: ShaleStore<Node> + Send + Sync, T> Stream for MerkleKeyValueStream<'
     }
 }
 
+enum PathIteratorState<'a> {
+    Iterating {
+        /// The key, as nibbles, of the node at [address], without the
+        /// node's partial path (if any) at the end.
+        /// Invariant: If this node has a parent, the parent's key is a
+        /// prefix of the key we're seeking.
+        /// Note the node at [address] may not have a key which is a
+        /// prefix of the key we're seeking.
+        matched_key: Vec<u8>,
+        unmatched_key: NibblesIterator<'a, 0>,
+        address: DiskAddress,
+    },
+    Exhausted,
+}
+
+/// Iterates over all nodes on the path to a given key starting from the root.
+/// If the key isn't in the trie, the final returned node is either the node with the
+/// longest prefix of the key or the node that is where the key would be inserted if
+/// it were in the trie. Note this means the final returned node may not be a prefix
+/// of the key we're seeking.
+pub struct PathIterator<'a, 'b, S, T> {
+    state: PathIteratorState<'b>,
+    merkle: &'a Merkle<S, T>,
+}
+
+impl<'a, 'b, S: ShaleStore<Node> + Send + Sync, T> PathIterator<'a, 'b, S, T> {
+    pub(super) fn new(
+        key: &'b [u8],
+        merkle: &'a Merkle<S, T>,
+        merkle_root: DiskAddress,
+    ) -> Result<Self, MerkleError> {
+        let sentinel_node = merkle.get_node(merkle_root)?;
+
+        let root = match sentinel_node.inner() {
+            NodeType::Branch(branch) => match branch.children[0] {
+                Some(root) => root,
+                None => {
+                    return Ok(Self {
+                        state: PathIteratorState::Exhausted,
+                        merkle,
+                    })
+                }
+            },
+            _ => unreachable!("sentinel node is not a branch"),
+        };
+
+        Ok(Self {
+            merkle,
+            state: PathIteratorState::Iterating {
+                matched_key: vec![],
+                unmatched_key: Nibbles::new(key).into_iter(),
+                address: root,
+            },
+        })
+    }
+}
+
+impl<'a, 'b, S: ShaleStore<Node> + Send + Sync, T> Iterator for PathIterator<'a, 'b, S, T> {
+    type Item = Result<(Key, NodeObjRef<'a>), MerkleError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        // destructuring is necessary here because we need mutable access to `state`
+        // at the same time as immutable access to `merkle`.
+        let Self { state, merkle } = &mut *self;
+
+        match state {
+            PathIteratorState::Exhausted => None,
+            PathIteratorState::Iterating {
+                matched_key,
+                unmatched_key,
+                address,
+            } => {
+                let node = match merkle.get_node(*address) {
+                    Ok(node) => node,
+                    Err(e) => return Some(Err(e)),
```

Comment:
Saves you ONE WHOLE CHARACTER!!! Haha, not sure if you knew you could do this, that's the only reason the suggestion is here, take it or leave it. 
```suggestion
                    e @ Err(_) => return Some(e),
```